pkg/sql/queuefeed/reader.go

```diff
@@ -240,39 +281,88 @@ func (r *Reader) GetRows(ctx context.Context, limit int) ([]tree.Datums, error)
 		}
 	}
 
-	if limit > len(r.mu.buf) {
-		limit = len(r.mu.buf)
+	fmt.Printf("GetRows called with limit: %d, buf len: %d, numDataEvents: %d\n", limit, len(r.mu.buf), r.mu.numDataEvents)
+
+	// Iterate through buffer, collecting data events and tracking checkpoints
+	position := 0
+	eventCount := 0
+	for eventCount < limit && position < len(r.mu.buf) {
+		event := r.mu.buf[position]
+		position++
+
+		if !event.resolved.IsEmpty() {
+			// This is a checkpoint event
+			r.mu.inflightCheckpoint = event.resolved
+			fmt.Printf("GetRows: found checkpoint at position %d: %s\n", position-1, event.resolved)
+		} else {
+			// This is a data event
+			r.mu.inflightBuffer = append(r.mu.inflightBuffer, event.row)
+			eventCount++
```

Comment:
why not include the resolved events in the inflight buffer? the copy would be cheaper probably and it would be slightly simpler